### DISCUSSION TITLE
The ability to click on GLTF decorations even when not using the select tool

### DISCUSSION THREAD

**Discussion by salah-walid** [iTwin/itwinjs-core#4929] [Q&A] [answered]
Hi ITwin

Using [this gltf decoration](https://www.itwinjs.org/sandboxes/NathanElrod/glTF%20Decorator) and adding `onDecorationButtonEvent` we are able to handle a click event on the GLTF model.
The problem comes when trying to click on it when having other tools then the select tool, a problem that doesn't seem to happen when using a 2D marker using the [Marker](https://www.itwinjs.org/reference/core-frontend/views/marker/) class.
Is there a way to have the desired effect when using the Decorator interface (IE clicking and receiving the said click event on the GLTF Model when the selection tool is not active) ?

**Comment by bbastings** [comment]
Marker locate is relatively simple, it's strictly a check if the view coordinates of the cursor are inside the marker's size rectangle (in pixels). A quick point inside rectangle check is made before deciding to pass the event on to the active tool or to the marker.

Locating decorations is more involved as the locate is based on the decoration's graphics. Locate here works off the depth buffer, just like it does for a persistent element's graphics. As such, picking decorations requires the cooperation of the active tool. First the tool needs to be one that enables locate (viewing tools and tools that just enable AccuSnap do not enable locate). The tool then also needs to explicitly enable the locate of decoration graphics by setting IModelApp.locateManager.options.allowDecorations to true.

For a tool that you write, you can choose to allow decorations to be selected just like the SelectTool does. HTH

**Reply by salah-walid** [reply]
How to enable locate on the primitive tool for example ?

**Comment by bbastings** [comment]
You'll want to call [initLocateElements](https://www.itwinjs.org/reference/core-frontend/tools/interactivetool/initlocateelements/) in your tool. For a tool that immediately wants to start locating elements you can do something like this:

  public override async onPostInstall() {
    await super.onPostInstall();
    this.initLocateElements();
    IModelApp.locateManager.options.allowDecorations = true; // Allow decorations
  }

You can also look at [SelectionTool](https://www.itwinjs.org/reference/core-frontend/selectionset/selectiontool/) for everything it does.

**Reply by salah-walid** [reply]
We are using the primitive tool for navigation(left click rotate, right click pan), is there a way to override the primitive tool to allow the locate feature ?

**Comment by bbastings** [comment]
Not sure what "primitive tool" you are referring too?

Are you maybe talking about the default IdleTool behavior after setting ToolAdmin.defaultToolId to not specify a default tool?

**Reply by salah-walid** [reply]
Using [This sample](https://www.itwinjs.org/sandboxes/JoshSchifter/Right-Mouse-Pan) to make left drag rotate and right drag pan, which has the effect of making the default tool not be the select tool, thus not being able to capture GLTF click event, is there a way to be able to do that when using this hack ?

**Reply by bbastings** [reply]
While it is possible to supply your own IdleTool for further customization as mentioned in that sample, enabling locate for the IdleTool would not be an appropriate solution.

The reason for this is that the IdleTool is not installed and is never the active tool, it's just there to handle some events when there is either  no active tool, or the active tool does not choose to handle the event itself. To correctly support flashing the element under the cursor, locate tooltips, etc. you really want to be doing this in the active tool. Doing so allows your locate settings to be properly suspended by viewing tools and properly initialized only when your tool is active.

You should create a simple sub-class of PrimitiveTool that you set as the default tool for your app. In this PrimitiveTool you would enable locate and call the decoration button event method from an override of onDataButtonDown as per SelectionTool. If a decoration is not clicked on you would start the appropriate viewing tool as per IdleTool.

**Reply by salah-walid** [reply]
I have tried with this combination and it doesn't work :
```Typescript
export default class CustomPrimitiveTool extends PrimitiveTool {
  public static toolId = "Custom.PrimitiveTool";

  onRestartTool(): Promise<void> {
    return this.exitTool();
  }

  public override async onPostInstall() {
    this.initLocateElements(true);
    IModelApp.locateManager.options.allowDecorations = true; // Allow decorations
    return await super.onPostInstall();
  }

  async onDataButtonDown(_ev: BeButtonEvent): Promise<EventHandled> {
    this.initLocateElements(true);
    IModelApp.locateManager.options.allowDecorations = true; // Allow decorations
    return EventHandled.Yes
  }
}
```

```Typescript
const onAppInit = useCallback(
    () => {
      void IModelApp.localization.registerNamespace("marker-pin-i18n-evercam");
      CustomPrimitiveTool.register("marker-pin-i18n-evercam");
      IModelApp.tools.run(CustomPrimitiveTool.toolId);
      IModelApp.toolAdmin.defaultToolId = CustomPrimitiveTool.toolId;
      FrontstageManager.onFrontstageActivatedEvent.addListener((args: FrontstageActivatedEventArgs) => {
        // clear out default tool getter so default tool is not set
        Object.defineProperty(args.activatedFrontstageDef, "_defaultTool", {value: CustomPrimitiveTool.toolId});
        IModelApp?.toolAdmin?.startDefaultTool();
      })
    },
    [],
)


return <Viewer
            iTwinId={iTwinId}
            iModelId={iModelId}
            authClient={myTokenServerAuthClient}
            enablePerformanceMonitors={false}
            viewCreatorOptions={{ viewportConfigurer: (_v) => {} }}
            mapLayerOptions={AppStoreInstance.viewerType === "site-view" ? mapLayerOptions : {}}
            tileAdminOptions={{ cesiumIonKey: `${cesiumKey}` }}
            onIModelAppInit={onAppInit}
          />
```

Can you please provide an example where we can use the primitive tool but still have locate

**Reply by bbastings** [reply]
The goal is to create a tool that combines the logic from SelectTool for a left click on a decoration with the left/right drag logic for starting viewing tools from IdleTool. Something like this:

```
export class ClickDecorationTool extends PrimitiveTool {
  public static override toolId = "ClickDecoration";
  public static override iconSpec = "icon-cursor";

  public override requireWriteableTarget(): boolean { return false; }
  public override autoLockTarget(): void { }

  public override async onDataButtonUp(ev: BeButtonEvent): Promise<EventHandled> {
    if (!ev.viewport)
      return EventHandled.No;

    const hit = await IModelApp.locateManager.doLocate(new LocateResponse(), true, ev.point, ev.viewport, ev.inputSource);
    if (hit && !(hit.isModelHit || hit.isMapHit || hit.isElementHit))
      await IModelApp.viewManager.onDecorationButtonEvent(hit, ev);

    return EventHandled.No;
  }

  public override async onMouseStartDrag(ev: BeButtonEvent): Promise<EventHandled> {
    if (!ev.viewport)
      return EventHandled.No;

    let toolId: string;

    switch (ev.button) {
      case BeButton.Data:
        toolId = "View.Rotate";
        break;

      case BeButton.Reset:
        toolId = "View.Pan";
        break;

      default:
        return EventHandled.No;
    }

    const viewTool = IModelApp.tools.create(toolId, ev.viewport, true, true) as ViewManip | undefined;
    if (viewTool && await viewTool.run())
      return viewTool.startHandleDrag(ev);
    return EventHandled.Yes;
  }

  public async onRestartTool(): Promise<void> { return this.exitTool(); }

  public override async onPostInstall() {
    await super.onPostInstall();
    this.initLocateElements(true);
    IModelApp.locateManager.options.allowDecorations = true; // Allow decorations
  }
}
```

**Reply by salah-walid** [reply]
Thanks, that solved my issue, one last question is there a way to enable IModel component selection, to show it's properties on the properties panel ?

**Comment by bbastings** [comment, marked as answer]
The properties panel works off selection set changed event. If you want to select elements you should consider just using the default SelectTool and getting used to panning with the middle mouse button and rotating with shift+middle mouse button...

**Reply by bbastings** [reply]
You can try this (I did not test), sub-class from SelectionTool and override onMouseStartDrag to replace being able to use drag box and crossing line selection with pan and rotate. You will still be able to do box and crossing line selection by changing the selection method tool setting.

```
class PanRotateSelectionTool extends SelectionTool {
  public static override toolId = "PanRotateSelection";

  public override async onMouseStartDrag(ev: BeButtonEvent): Promise<EventHandled> {
    if (!ev.viewport)
      return EventHandled.No;

    let toolId: string;

    switch (ev.button) {
      case BeButton.Data:
        toolId = "View.Rotate";
        break;

      case BeButton.Reset:
        toolId = "View.Pan";
        break;

      default:
        return super.onMouseStartDrag(ev);
    }

    const viewTool = IModelApp.tools.create(toolId, ev.viewport, true, true) as ViewManip | undefined;
    if (viewTool && await viewTool.run())
      return viewTool.startHandleDrag(ev);
    return EventHandled.Yes;
  }
}
```

**Reply by salah-walid** [reply]
Thank you @bbastings, You have been of great help.
